File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/25-SingleSourceShortestPath/LeetCode/LeetCode-1514-2.py
class Solution:
    # Bellman-Ford algorithm. Gets TLE
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start: int, end: int) -> float:
        cost = [0]*n
        cost[start] = 1.0

        for iter in range(0, n-1):
            for i in range(len(edges)):
                u = edges[i][0]
                v = edges[i][1]
                p = succProb[i]
                cost[v] = max(cost[v], cost[u]*p)  # Relax(u, v)
                cost[u] = max(cost[u], cost[v]*p)  # Relax(v, u)
        return cost[end]

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/25-SingleSourceShortestPath/LeetCode/LeetCode-1514-2.cpp
class Solution {
public:
    /// Bellman-Ford algorithm. Gets TLE
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start, int end) {       
        vector<double> cost(n, 0);            
        cost[start] = 1.0;

        for (int iter=0; iter<n-1; iter++){
            for (int i=0; i<edges.size(); i++){
                int u = edges[i][0];
                int v = edges[i][1];
                double p = succProb[i];                
                cost[v] = std::max(cost[v], cost[u]*p); // Relax(u, v)
                cost[u] = std::max(cost[u], cost[v]*p); // Relax(v, u)
            } // end-for         
        } // end-for

        return cost[end];
    } // end-maxProbability
};